settings: Fix text overflow while hovering over reactions/Users hover/viewing profile card.

SahilSingh177 opened the pull request:
Currently, if a reaction has been made by a user with a long name, the text overflows when hovering over the reaction.

The same bug appears while hovering over long usernames in user's profile card or in the Users section.

To fix this, we add a span tag to the usernames and assign it a class 'break_all_words' and apply 'word-break:break-all' property to the class to prevent overflow.

To handle overflow in case of user's profile or in the users section, we add a template that handles the username tooltip for the user's profile card add class 'break_all_words' to the span element that contains the username.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

Previous behavior: 

![reaction_bug](https://user-images.githubusercontent.com/96344003/218385667-3ac09538-b130-47ad-9840-9d5c8020b2e3.png)

<img src="https://user-images.githubusercontent.com/96344003/225551519-fbf78efd-483a-403e-8869-7cd0d020f062.png" width="600" height="300"/>

<img src="https://user-images.githubusercontent.com/96344003/225553033-b8fea879-42cc-49cc-b718-83bb56f81983.png" width="600" height="400"/>

New behavior:

![reaction_hover_fix](https://user-images.githubusercontent.com/96344003/224756165-fb1ba8d9-8cfa-428a-91f8-0483446b90fc.png)

<img src="https://user-images.githubusercontent.com/96344003/225552047-d5bcb20f-e972-44ec-a5cf-230d439acc85.png" width="600" height="400"/>

<img src="https://user-images.githubusercontent.com/96344003/225552055-66f556e8-694a-4e58-985c-87cf599b95c8.png" width="500" height="300"/>

<img src="https://user-images.githubusercontent.com/96344003/225552840-88d97243-145d-4e74-af93-92d41c661d5f.png" width="600" height="300"/>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
This change will apply to ALL tooltips, so we should be pretty careful in testing this.

amanagr commented:
czo discussion: https://chat.zulip.org/#narrow/stream/9-issues/topic/Text.20exceeding.20the.20boundaries.20while.20hovering.20over.20reactions.2E

amanagr commented:
@SahilSingh177 this is not the right fix, we should wrap the sender names inside a `span` element with some class and apply `break-all` to that.

SahilSingh177 commented:
@amanagr, I'm not exactly sure how to wrap the user name inside a span element. 
However, I do have another idea that could work. We could add a class to the reaction tooltip message and apply the CSS property 'break-all' to that class. This would prevent the tooltip from overflowing and affecting other tooltips on the page.

Let me know what you think of this solution!

amanagr commented:
The difference being what I said and what you suggested is the use of `span` element around the user names which ensure that `break-all` only applies to the user name text and not to other parts of the message so that they remain intact and are understandable. To achieve this, you need to modify `reactions.generate_title` function.

SahilSingh177 commented:
@amanagr, I have made changes as per your suggestion. Please review and let me know if you require any further changes. Thank you!

amanagr commented on web/src/reactions.js:
```diff
@@ -154,18 +154,24 @@ function generate_title(emoji_name, user_ids) {
         }
         context.username = usernames[0];
-        return $t({defaultMessage: "{username} reacted with {emoji_name}"}, context);
+        return $t_html(
+            {defaultMessage: "<div><span>{username}</span> reacted with {emoji_name}</div>"},
+            context,
+        );
```
@SahilSingh177 Looks great! Do we need the `div` here for some reason?

SahilSingh177 replied:
Yes, the ```div``` element acts as a container, enclosing the ```span``` elements and preventing them from appearing side-by-side, which would happen if the ```div``` element was not used.
![hover-bug](https://user-images.githubusercontent.com/96344003/224799442-3b64e3f4-2807-471d-9dc2-1461f2a3438f.png)

alya commented:
Hm, I'm still seeing overflow in the right sidebar tooltip when I test this.

<img width="548" alt="Screen Shot 2023-03-13 at 10 58 08 PM" src="https://user-images.githubusercontent.com/2090066/224909550-c0dbc814-3e22-4895-ad3a-969ee827b92b.png">

SahilSingh177 commented:
> Hm, I'm still seeing overflow in the right sidebar tooltip when I test this.
> 
> <img alt="Screen Shot 2023-03-13 at 10 58 08 PM" width="548" src="https://user-images.githubusercontent.com/2090066/224909550-c0dbc814-3e22-4895-ad3a-969ee827b92b.png">

The changes that I made in this PR affects only the ```reaction-tooltip``` and not others.
I mentioned this issue in [CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/Text.20exceeding.20the.20boundaries.20while.20hovering.20over.20reactions.2E/near/1523096).

Should I open another PR for this?

amanagr commented:
It is fine to include it in this PR in a separate commit.

SahilSingh177 commented on web/src/tippyjs.js:
```diff
@@ -105,6 +105,19 @@ export function initialize() {
         target: ".tippy-zulip-tooltip",
     });
 
+    delegate("body", {
```
Added a separate tooltip configuration to manage the tooltip displayed when hovering over a User's profile card. Previously, it used `data-tippy-content`, so applying CSS to it could potentially break other tooltips.

SahilSingh177 commented on web/src/tippyjs.js:
```diff
@@ -105,6 +105,19 @@ export function initialize() {
+        },
+        appendTo: () => document.body,
+    });
+
```
Added a separate tooltip configuration to manage the tooltip displayed when hovering over a User's profile card. Previously, it used `data-tippy-content`, so applying CSS to it could potentially break other tooltips.

amanagr commented on web/src/tippyjs.js:
```diff
@@ -105,6 +105,19 @@ export function initialize() {
+        placement: "top",
+        onShow(instance) {
+            instance.setContent(instance.reference.textContent);
+            instance.popper.querySelector(".tippy-content").classList.add("tooltip_username");
```
We can create a handlebar which takes `class_name` as context and use it since this can be slow to take effect which can cause user to see a transition state where the css is not applied to being applied.

SahilSingh177 replied:
Are you suggesting something like this?
```
const context = {
    username: instance.reference.textContent,
    class_name:'tooltip_username'
};

instance.setContent(parse_html(
  $t_html(
    {defaultMessage: "<span>{username}</span>"},
    context,
  )
));
```

The issue with implementing this approach is that including `class='{class_name}'` within the `span` tag results in a parsing error. Is there a viable solution available to resolve this problem?

SahilSingh177 replied:
Are you suggesting something like this?
```
const context = {
    username: instance.reference.textContent,
    class_name:'tooltip_username'
};

instance.setContent(parse_html(
  $t_html(
    {defaultMessage: "<span>{username}</span>"},
    context,
  )
));
```

The issue with implementing this approach is that including `class='{class_name}'` within the `span` tag results in a parsing error. Is there a viable solution available to resolve this problem?

amanagr replied:
I meant we can create a template like `message_inline_image_tooltip`

amanagr commented on web/src/tippyjs.js:
```diff
@@ -126,9 +139,9 @@ export function initialize() {
+                instance.setContent(parse_html(title));
             }
-
+            instance.popper.querySelector(".tippy-content").classList.add("reaction-title");
```
We can use the same class name for both.

amanagr commented:
@SahilSingh177 posted a few comments, looks great!

SahilSingh177 commented:
@amanagr, can you please have a look at the code and suggest what might be causing the CodeQL test to fail. I tried to figure it out, but couldn't find any solution.

SahilSingh177 commented:
@amanagr, I have fixed the failing check and updated the PR with the changes you suggested. Could you please review them and let me know if any further changes are required?
Thank you!

amanagr commented:
@SahilSingh177 I pushed a change that I think would work best, can you review it and split the commit for reactions and usernames separately?

SahilSingh177 commented:
@amanagr, I have split the commits and can confirm that the code works perfectly.

amanagr commented:
LGTM! Thanks!

SahilSingh177 committed:
reactions: Fix text-overflow while hovering over reactions.

Currently, if reaction has been made by a user
with a long name, the text overflows while
hovering over the reaction.

We fix this by adding a span tag to the usernames
and assigning it a class 'break_all_words' and
applying 'word-break:break-all' property to
the class to prevent overflow.

SahilSingh177 committed:
tippy: Fix text overflow on Users hover/viewing profile card.

Currently, text overflows on hovering over
long usernames in user's profile card or
in the Users section.

To fix this, we add a template that handles the
username tooltip for the user's profile card and
add class 'break_all_words' to the span element
that contains the username in the user's profile
card and the tooltip that appears while hovering
over the users section.

timabbott commented on web/src/reactions.js:
```diff
@@ -184,18 +190,18 @@ function generate_title(emoji_name, user_ids) {
         {
             defaultMessage:
-                "{comma_separated_usernames} and {last_username} reacted with {emoji_name}",
+                "<div><span>{comma_separated_usernames}</span> and <span>{last_username}</span> reacted with {emoji_name}</div>",
```
I don't like adding this much HTML into the translation tag blocks.  The `<div>` definitely doesn't need to part of the translation block. I think the best option might be to move these constructions to a handlebars template? I'm not sure.

I'm also not really sure this change as written makes sense -- `comma_separated_usernames` contains multiple usernames anyway, so you're not enclosing each one on its own span regardless.

SahilSingh177 replied:
Yes, I also think that it'll make the code more readable. I'll update the PR with the suggested changes.

timabbott commented on web/src/tippyjs.js:
```diff
@@ -130,7 +130,10 @@ export function initialize() {
+                instance.setContent(parse_html(title));
+                instance.popper
+                    .querySelector(".tippy-content > div > span")
+                    .classList.add("break_all_words");
```
Rather than adding the CSS class here, we should be adding it when we generate the actual spans, I think; it's way easier to understand that way.
